compiler/src/dotty/tools/dotc/quoted/PickledQuotes.scala

```diff
@@ -68,16 +68,19 @@ object PickledQuotes {
   }
 
   /** Replace all term holes with the spliced terms */
-  private def spliceTerms(tree: Tree, typeHole: (Int, Seq[Any]) => scala.quoted.Type[?], termHole: (Int, Seq[Any], scala.quoted.Quotes) => scala.quoted.Expr[?])(using Context): Tree = {
+  private def spliceTerms(tree: Tree, typeHole: (Int, Seq[Any]) => scala.quoted.Type[?], termHole: (Int, Seq[Any], scala.quoted.Quotes) => scala.quoted.Expr[?], version: Int)(using Context): Tree = {
     val evaluateHoles = new TreeMap {
       override def transform(tree: tpd.Tree)(using Context): tpd.Tree = tree match {
-        case Hole(isTerm, idx, args) =>
+        case Hole(isTermHole, idx, args, _, _) =>
           inContext(SpliceScope.contextWithNewSpliceScope(tree.sourcePos)) {
             val reifiedArgs = args.map { arg =>
-              if (arg.isTerm) (q: Quotes) ?=> new ExprImpl(arg, SpliceScope.getCurrent)
+              if (arg.isTerm)
+                version match
+                  case 1 => (q: Quotes) ?=> new ExprImpl(arg, SpliceScope.getCurrent)
+                  case 2 => new ExprImpl(arg, SpliceScope.getCurrent)
```

Comment:
It seems the lambdas here are called during compilation. What will happen if we don't do the versioning? And do well need the versioning at the call-site of `reifiedArgs`?

Reply:
> What will happen if we don't do the versioning?

Then macros that were compiled with a previous version of the compiler would not work with this version of the compiler. With this versioning, we can keep supporting any macro compiled since `3.0.0`.

> And do well need the versioning at the call-site of `reifiedArgs`?

Not sure if I understand correctly what you mean by _call-site of `reifiedArgs`_. If that refers to the generated code that is passed as `termHole` then that change is what is being reflected here.

Reply:
> Not sure if I understand correctly what you mean by call-site of reifiedArgs. If that refers to the generated code that is passed as termHole then that change is what is being reflected here.

I may be asking a stupid question here: Here we create a value based on `version`, when the value is used (elsewhere), do we need to handle two different values corresponding to the two `version`s at the usage site?

I think you already hinted at the answer above: the usage site for `version = 1` is from 3.0.0 macros, while values corresponding to `version = 2` will be consumed by macros compiled by this PR.

Reply:
> I think you already hinted at the answer above: the usage site for version = 1 is from 3.0.0 macros, while values corresponding to version = 2 will be consumed by macros compiled by this PR.

That is right. For macros compiled from 3.0.0 up to this PR the compiler-generated code that called `unpickleExpr` which we internally handle as `version = 1`. This code expected the hole arguments to be passed as `Quotes ?=> Expr[Any]`. From the PR onwards the compiler will generate a call `unpickleExprV2` which is internally handled as `version = 2` which handles the hole arguments as `Expr[Any]`. Therefore at use-site we know the version based on which version of `unpickleExpr`/`unpickleExprV2` we call and the compiler is capable of handling both of them.

Of course, an old compiler will not be able to execute a macro compiled by this version of the compiler.

Another design option for versioning was to add a `unpickleExprVersioned` that takes a parameter the version number. `unpickleExpr` would just call `unpickleExprVersioned` with `version = 1`. This has the drawback of adding this extra parameter everywhere which seemed unnecessary overhead. Furthermore, if we have a next version we might change the signature of `unpickleExpr` completely. For example, encoding the arguments in a different way, maybe to optimize some common cases.

Reply:
> Of course, an old compiler will not be able to execute a macro compiled by this version of the compiler.

What is the impact of this? Does it mean it has to be in a bigger compiler version, such as `3.2`?

Reply:
A macro compiled with 3.2+ cannot be used for code compiled with 3.0 or 3.1